File: ch02/2.6/java_src/Lexer.java
package lexer;      //File Lexer.java

import java.io.*;
import java.util.*;

public class Lexer
{
    public int line = 1;
    private char peek = ' ';
    private Hashtable words = new Hashtable();
    void reserve(Word t)
    {
        words.put(t.lexme, t);
    }

    public Lexer()
    {
        reserve(new Word(Tag.TRUE, "true"));
        reserve(new Word(Tag.FALSE, "false"));
    }

    public Token scan() throws IOException
    {
        for( ; ; peek = (char)System.in.read())
        {
            if( peek == ' ' || peek == '\t')
                continue;
            else if (peek == '\n') 
                ++line;
            else break;
        }

        if( Character.isDigit(peek))
        {
            int v = 0;
             do{
                v = 10 * v + Character.digit(peek, 10);
                peek = (char)System.in.read();
             }while( Character.isDigit(peek));

             return new Num(v);
        }

        if( Character.isLetter(peek))
        {
            StringBuffer b = new StringBuffer();
            do{
                b.append(peek);
                peek = (char)System.in.read();
            } while(Character.isLetterOrDigit(peek));

            String s = b.toString();
            Word w = (Word)words.get(s);
            if(w != null)
                return w;
            w = new Word(Tag.ID, s);
            words.put(s,w);
            
            return w;
        }

        Token t = new Token(peek);
        peek = ' ';
        return t;
    }
}

File: ch02/2.6/cpp_src/lexer/lexer.cpp
#include "lexer.h"
#include <iostream>

namespace dragon {namespace ch2 {

TokenSptr Lexer::scan()
{
    for(; ; std::cin >> peek)
    {
        if( peek == ' ' || peek == '\t')
            continue;
        else if (peek == '\n')
            ++curr_line;
        else break;
    }

    if(std::isdigit(peek))
    {
        int v = 0;
        do{
            v = v*10 + peek - 48;
            std::cin >> peek;
        }while(std::isdigit(peek));

        return std::make_shared<Num>(v);
    }

    if(std::isalpha(peek))
    {
        std::string key{};
        do{
            key.push_back(peek);
            std::cin >> peek;
        }while(std::isalpha(peek)   ||  std::isdigit(peek));

        auto w = words.find(key);
        if(w != words.end())    return w->second;
        auto ret = std::make_shared<Word>(Tag::ID, key);
        words.insert({key, ret});
        return ret;
    }
    TokenSptr token = std::make_shared<Token>(Token{peek});
    peek = ' ';

    return  token;
}

}}//namespace


